File: engine/language_client_ruby/test/test_cancellation.rb
# frozen_string_literal: true

require "test/unit"
require "sorbet-runtime"
require_relative "../lib/stream"

# Mock FFI stream for testing
class MockFFIStream
  extend T::Sig

  def initialize
    @cancelled = false
    @mutex = Mutex.new
  end

  sig { void }
  def cancel
    @mutex.synchronize { @cancelled = true }
  end

  sig { returns(T::Boolean) }
  def is_cancelled
    @mutex.synchronize { @cancelled }
  end

  sig { params(callback: T.nilable(Proc)).void }
  def on_event(callback)
    @event_callback = callback
  end

  sig { params(ctx_manager: T.untyped).returns(T.untyped) }
  def done(ctx_manager)
    raise "Stream was cancelled" if is_cancelled
    
    # Simulate some partial events
    if @event_callback
      3.times do |i|
        break if is_cancelled
        result = OpenStruct.new(parsed: "partial_#{i}")
        @event_callback.call(result)
        sleep(0.01)
      end
    end
    
    OpenStruct.new(parsed: "final_result")
  end
end

class TestBamlStreamCancellation < Test::Unit::TestCase
  def setup
    @mock_ffi_stream = MockFFIStream.new
    @mock_ctx_manager = Object.new
    @stream = Baml::BamlStream.new(
      ffi_stream: @mock_ffi_stream,
      ctx_manager: @mock_ctx_manager
    )
  end

  def test_cancel_method_exists
    assert_respond_to(@stream, :cancel)
  end

  def test_cancelled_predicate_exists
    assert_respond_to(@stream, :cancelled?)
  end

  def test_initial_state_not_cancelled
    assert_false(@stream.cancelled?)
  end

  def test_cancel_sets_cancelled_state
    @stream.cancel
    assert_true(@stream.cancelled?)
  end

  def test_cancel_calls_ffi_stream_cancel
    @stream.cancel
    assert_true(@mock_ffi_stream.is_cancelled)
  end

  def test_cancelled_predicate_checks_ffi_stream
    @mock_ffi_stream.cancel
    assert_true(@stream.cancelled?)
  end

  def test_get_final_response_success
    result = @stream.get_final_response
    assert_equal("final_result", result)
  end

  def test_get_final_response_with_cancellation
    @stream.cancel
    
    assert_raises(RuntimeError, /cancelled/) do
      @stream.get_final_response
    end
  end

  def test_get_final_response_caches_result
    result1 = @stream.get_final_response
    result2 = @stream.get_final_response
    
    assert_equal(result1, result2)
    assert_same(result1, result2)
  end

  def test_each_iteration_with_cancellation
    partial_results = []
    
    # Cancel after short delay
    cancel_thread = Thread.new do
      sleep(0.05)
      @stream.cancel
    end
    
    @stream.each do |partial|
      partial_results << partial
      break if partial_results.length >= 5  # Safety limit
    end
    
    cancel_thread.join
    
    # Should have collected some partial results
    assert_operator(partial_results.length, :>=, 1)
    assert_true(@stream.cancelled?)
  end

  def test_each_without_block_returns_enumerator
    enumerator = @stream.each
    assert_instance_of(Enumerator, enumerator)
  end

  def test_multiple_cancellations_safe
    @stream.cancel
    @stream.cancel  # Should not raise error
    @stream.cancel  # Should not raise error
    
    assert_true(@stream.cancelled?)
  end

  def test_thread_safety_of_cancellation
    threads = []
    
    # Cancel from multiple threads simultaneously
    10.times do
      threads << Thread.new { @stream.cancel }
    end
    
    threads.each(&:join)
    
    assert_true(@stream.cancelled?)
  end

  def test_cancellation_prevents_resource_waste
    start_time = Time.now
    
    # Cancel immediately
    @stream.cancel
    
    assert_raises(RuntimeError, /cancelled/) do
      @stream.get_final_response
    end
    
    elapsed = Time.now - start_time
    
    # Should complete quickly due to immediate cancellation
    assert_operator(elapsed, :<, 0.1, "Cancellation should be immediate")
  end

  def test_cancellation_during_iteration
    partial_count = 0
    cancelled_during_iteration = false
    
    # Start iteration and cancel during it
    cancel_thread = Thread.new do
      sleep(0.02)  # Let iteration start
      @stream.cancel
    end
    
    begin
      @stream.each do |partial|
        partial_count += 1
        cancelled_during_iteration = @stream.cancelled?
        break if cancelled_during_iteration || partial_count >= 10
        sleep(0.01)  # Simulate processing time
      end
    rescue => e
      # Cancellation might cause iteration to stop with error
    end
    
    cancel_thread.join
    
    assert_true(@stream.cancelled?)
    # Should have detected cancellation during iteration
    assert_true(cancelled_during_iteration || partial_count == 0)
  end

  def test_coerce_methods_called
    # Test that subclasses can override coerce methods
    custom_stream = Class.new(Baml::BamlStream) do
      def coerce_partial(result)
        "partial_#{result}"
      end
      
      def coerce_final(result)
        "final_#{result}"
      end
    end.new(
      ffi_stream: @mock_ffi_stream,
      ctx_manager: @mock_ctx_manager
    )
    
    result = custom_stream.get_final_response
    assert_equal("final_final_result", result)
  end

  def test_error_handling_in_partial_processing
    # Mock FFI stream that causes errors in partial processing
    error_ffi_stream = Class.new(MockFFIStream) do
      def done(ctx_manager)
        if @event_callback
          # Simulate an error in partial processing
          result = OpenStruct.new(parsed: nil)  # This will cause coerce_partial to fail
          @event_callback.call(result)
        end
        
        OpenStruct.new(parsed: "final_result")
      end
    end.new
    
    error_stream = Baml::BamlStream.new(
      ffi_stream: error_ffi_stream,
      ctx_manager: @mock_ctx_manager
    )
    
    # Should handle errors gracefully and continue
    result = error_stream.get_final_response
    assert_equal("final_result", result)
  end

  def test_cleanup_on_cancellation
    partial_results = []
    
    @stream.cancel
    
    # Iteration should handle cancellation gracefully
    @stream.each do |partial|
      partial_results << partial
    end
    
    # Should not have processed any results due to cancellation
    assert_empty(partial_results)
  end
end

class TestBamlStreamIntegration < Test::Unit::TestCase
  def test_realistic_cancellation_scenario
    # Simulate a realistic scenario where user cancels a long-running operation
    mock_ffi_stream = MockFFIStream.new
    mock_ctx_manager = Object.new
    
    stream = Baml::BamlStream.new(
      ffi_stream: mock_ffi_stream,
      ctx_manager: mock_ctx_manager
    )
    
    start_time = Time.now
    results_collected = 0
    
    # Simulate user cancelling after seeing some results
    cancel_thread = Thread.new do
      sleep(0.03)  # Let some processing happen
      stream.cancel
    end
    
    begin
      stream.each do |partial|
        results_collected += 1
        break if results_collected >= 2  # User sees a couple results then cancels
      end
    rescue => e
      # Cancellation might cause errors
    end
    
    cancel_thread.join
    elapsed = Time.now - start_time
    
    assert_true(stream.cancelled?)
    assert_operator(elapsed, :<, 0.2, "Should complete quickly after cancellation")
  end

  def test_memory_cleanup_with_many_streams
    # Test that creating and cancelling many streams doesn't leak memory
    streams = []
    
    100.times do |i|
      mock_ffi_stream = MockFFIStream.new
      stream = Baml::BamlStream.new(
        ffi_stream: mock_ffi_stream,
        ctx_manager: Object.new
      )
      
      stream.cancel
      streams << stream
      
      assert_true(stream.cancelled?, "Stream #{i} should be cancelled")
    end
    
    # All streams should be cancelled
    assert_true(streams.all?(&:cancelled?))
  end
end

File: engine/language_client_python/python_src/tests/test_cancellation.py
"""
Comprehensive tests for BAML Python client cancellation functionality
"""

import asyncio
import pytest
import threading
import time
from unittest.mock import Mock, MagicMock

import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..'))

from baml_py.stream import BamlStream, BamlSyncStream


class TestBamlStreamCancellation:
    """Test async BamlStream cancellation functionality"""
    
    def setup_method(self):
        """Set up test fixtures"""
        self.mock_ffi_stream = Mock()
        self.mock_ffi_stream.cancel = Mock()
        self.mock_ffi_stream.is_cancelled = Mock(return_value=False)
        self.mock_ffi_stream.on_event = Mock()
        self.mock_ffi_stream.done = Mock()
        
        self.mock_ctx_manager = Mock()
        
        def partial_coerce(result):
            return {"partial": result}
        
        def final_coerce(result):
            return {"final": result}
        
        self.stream = BamlStream(
            self.mock_ffi_stream,
            partial_coerce,
            final_coerce,
            self.mock_ctx_manager
        )
    
    def test_cancel_method_exists(self):
        """Test that cancel method exists and is callable"""
        assert hasattr(self.stream, 'cancel')
        assert callable(self.stream.cancel)
    
    def test_cancel_calls_ffi_stream_cancel(self):
        """Test that cancel() calls the FFI stream's cancel method"""
        self.stream.cancel()
        
        self.mock_ffi_stream.cancel.assert_called_once()
        assert self.stream._cancelled is True
    
    def test_is_cancelled_method(self):
        """Test is_cancelled method functionality"""
        # Initially not cancelled
        assert not self.stream.is_cancelled()
        
        # After cancellation
        self.stream.cancel()
        assert self.stream.is_cancelled()
    
    def test_is_cancelled_checks_ffi_stream(self):
        """Test that is_cancelled also checks FFI stream status"""
        self.mock_ffi_stream.is_cancelled.return_value = True
        
        assert self.stream.is_cancelled()
        self.mock_ffi_stream.is_cancelled.assert_called()
    
    @pytest.mark.asyncio
    async def test_get_final_response_with_cancellation(self):
        """Test get_final_response raises error when cancelled"""
        self.stream.cancel()
        
        # Mock the FFI stream to raise an exception
        self.mock_ffi_stream.done.side_effect = Exception("Stream cancelled")
        
        with pytest.raises(RuntimeError, match="Stream was cancelled"):
            await self.stream.get_final_response()
    
    @pytest.mark.asyncio
    async def test_get_final_response_success(self):
        """Test get_final_response works when not cancelled"""
        mock_result = Mock()
        mock_result.parsed.return_value = {"message": "test"}
        self.mock_ffi_stream.done.return_value = mock_result
        
        result = await self.stream.get_final_response()
        
        assert result == {"final": {"message": "test"}}
        self.mock_ffi_stream.done.assert_called_once_with(self.mock_ctx_manager)
    
    @pytest.mark.asyncio
    async def test_async_iteration_with_cancellation(self):
        """Test async iteration stops when cancelled"""
        partial_results = []
        
        def mock_on_event(callback):
            # Simulate some partial results
            for i in range(3):
                if not self.stream.is_cancelled():
                    mock_result = Mock()
                    mock_result.parsed.return_value = {"step": i}
                    callback(mock_result)
        
        self.mock_ffi_stream.on_event.side_effect = mock_on_event
        
        # Mock final response to complete quickly
        mock_final = Mock()
        mock_final.parsed.return_value = {"final": "done"}
        self.mock_ffi_stream.done.return_value = mock_final
        
        # Cancel after short delay
        async def cancel_after_delay():
            await asyncio.sleep(0.05)
            self.stream.cancel()
        
        cancel_task = asyncio.create_task(cancel_after_delay())
        
        # Collect partial results
        async for partial in self.stream:
            partial_results.append(partial)
            if len(partial_results) >= 2:  # Limit to avoid infinite loop
                break
        
        await cancel_task
        
        # Should have collected some partial results before cancellation
        assert len(partial_results) >= 1
        assert all("partial" in result for result in partial_results)
    
    @pytest.mark.asyncio
    async def test_multiple_cancellations_safe(self):
        """Test that multiple cancel calls are safe"""
        self.stream.cancel()
        self.stream.cancel()  # Should not raise error
        self.stream.cancel()  # Should not raise error
        
        # FFI cancel should only be called once due to _cancelled flag
        self.mock_ffi_stream.cancel.assert_called_once()


class TestBamlSyncStreamCancellation:
    """Test sync BamlSyncStream cancellation functionality"""
    
    def setup_method(self):
        """Set up test fixtures"""
        self.mock_ffi_stream = Mock()
        self.mock_ffi_stream.cancel = Mock()
        self.mock_ffi_stream.is_cancelled = Mock(return_value=False)
        self.mock_ffi_stream.on_event = Mock()
        self.mock_ffi_stream.done = Mock()
        
        self.mock_ctx_manager = Mock()
        
        def partial_coerce(result):
            return {"partial": result}
        
        def final_coerce(result):
            return {"final": result}
        
        self.stream = BamlSyncStream(
            self.mock_ffi_stream,
            partial_coerce,
            final_coerce,
            self.mock_ctx_manager
        )
    
    def test_sync_cancel_method(self):
        """Test sync cancel method"""
        self.stream.cancel()
        
        self.mock_ffi_stream.cancel.assert_called_once()
        assert self.stream._cancelled is True
    
    def test_sync_is_cancelled_method(self):
        """Test sync is_cancelled method"""
        assert not self.stream.is_cancelled()
        
        self.stream.cancel()
        assert self.stream.is_cancelled()
    
    def test_sync_get_final_response_with_cancellation(self):
        """Test sync get_final_response with cancellation"""
        self.stream.cancel()
        
        self.mock_ffi_stream.done.side_effect = Exception("Stream cancelled")
        
        with pytest.raises(RuntimeError, match="Stream was cancelled"):
            self.stream.get_final_response()
    
    def test_sync_get_final_response_success(self):
        """Test sync get_final_response success"""
        mock_result = Mock()
        mock_result.parsed.return_value = {"message": "test"}
        self.mock_ffi_stream.done.return_value = mock_result
        
        result = self.stream.get_final_response()
        
        assert result == {"final": {"message": "test"}}
    
    def test_sync_iteration_with_cancellation(self):
        """Test sync iteration with cancellation"""
        partial_results = []
        
        def mock_on_event(callback):
            # Simulate partial results
            for i in range(5):
                if not self.stream.is_cancelled():
                    mock_result = Mock()
                    mock_result.parsed.return_value = {"step": i}
                    callback(mock_result)
                    time.sleep(0.01)  # Small delay
        
        self.mock_ffi_stream.on_event.side_effect = mock_on_event
        
        # Mock final response
        mock_final = Mock()
        mock_final.parsed.return_value = {"final": "done"}
        self.mock_ffi_stream.done.return_value = mock_final
        
        # Cancel after short delay in another thread
        def cancel_after_delay():
            time.sleep(0.05)
            self.stream.cancel()
        
        cancel_thread = threading.Thread(target=cancel_after_delay)
        cancel_thread.start()
        
        # Collect partial results
        for partial in self.stream:
            partial_results.append(partial)
            if len(partial_results) >= 3:  # Limit to avoid infinite loop
                break
        
        cancel_thread.join()
        
        # Should have collected some results
        assert len(partial_results) >= 1


class TestCancellationIntegration:
    """Integration tests for cancellation functionality"""
    
    @pytest.mark.asyncio
    async def test_cancellation_prevents_resource_waste(self):
        """Test that cancellation prevents resource waste"""
        mock_ffi_stream = Mock()
        mock_ffi_stream.cancel = Mock()
        mock_ffi_stream.is_cancelled = Mock(return_value=False)
        mock_ffi_stream.on_event = Mock()
        
        # Mock a slow operation
        async def slow_done(ctx_manager):
            await asyncio.sleep(1.0)  # Simulate slow operation
            return Mock()
        
        mock_ffi_stream.done = slow_done
        
        stream = BamlStream(
            mock_ffi_stream,
            lambda x: x,
            lambda x: x,
            Mock()
        )
        
        # Start the operation and cancel quickly
        start_time = time.time()
        
        async def quick_cancel():
            await asyncio.sleep(0.1)
            stream.cancel()
        
        cancel_task = asyncio.create_task(quick_cancel())
        
        with pytest.raises(RuntimeError, match="Stream was cancelled"):
            await stream.get_final_response()
        
        await cancel_task
        elapsed = time.time() - start_time
        
        # Should complete quickly due to cancellation
        assert elapsed < 0.5, f"Cancellation took too long: {elapsed}s"
        mock_ffi_stream.cancel.assert_called_once()
    
    def test_thread_safety_of_cancellation(self):
        """Test that cancellation is thread-safe"""
        mock_ffi_stream = Mock()
        mock_ffi_stream.cancel = Mock()
        mock_ffi_stream.is_cancelled = Mock(return_value=False)
        
        stream = BamlSyncStream(
            mock_ffi_stream,
            lambda x: x,
            lambda x: x,
            Mock()
        )
        
        # Cancel from multiple threads simultaneously
        def cancel_stream():
            stream.cancel()
        
        threads = []
        for _ in range(10):
            thread = threading.Thread(target=cancel_stream)
            threads.append(thread)
            thread.start()
        
        for thread in threads:
            thread.join()
        
        # Should be cancelled and FFI cancel should be called
        assert stream.is_cancelled()
        mock_ffi_stream.cancel.assert_called()


if __name__ == "__main__":
    pytest.main([__file__, "-v"])
